src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,117 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private T[] elements;
+
+    public ArrayList() {
+        this.elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        resize();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(index
+                    + " or less than 0 or greater than " + size);
+        }
+        if (size == index) {
+            add(value);
+        } else {
+            resize();
+            for (int i = size; i > index; i--) {
+                elements[i] = elements[i - 1];
+            }
+            elements[index] = value;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        errorChecking(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        errorChecking(index);
+        if (index == size) {
+            add(value);
+        } else {
+            elements[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        errorChecking(index);
+        T[] beginning = (T[]) new Object[index];
+        T[] end = (T[]) new Object[elements.length - index - 1];
```

Comment:
you don't need to create separate arrays for implementation of this method
you just need to shift elemnts after index to the left of elements array